File: day20/particle_swarm2.py
#!/usr/bin/env python3
from sys import argv
from Particle import *

part=[]
filename = argv[1]
with open(filename) as f:
    for line in f:
        part.append(Particle(line.rstrip()))

MAX_TICKS = 100 # worked empirically on my data
for t in range(1, MAX_TICKS+1):
    for p in part:
        p.tick()
    for i in range(0, len(part)-1):
        for j in range(i+1, len(part)):
            if part[i].collision(part[j]):
                part[i].set_hit()
                part[j].set_hit()

    part = [x for x in part if x.unhit()]
    print("t=", t, "remain=", len(part))

print("result2:", len(part))

    

File: day20/particle_swarm2.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);
use lib '.';
use Particle;

my @part;

while (<>) {
    chomp;
    push @part, Particle->new($_);
}

my $MAX_TICKS = 100; # worked empirically on my data
for my $t (1..$MAX_TICKS) {
    $_->tick() for @part;
    for my $i (0..$#part-1) {
        for my $j ($i+1..$#part) {
            if ($part[$i]->collision($part[$j])) {
                $part[$i]->set_hit(1);
                $part[$j]->set_hit(1);
            }
        }
    }
    @part = grep {$_->unhit()} @part;
    say "t=$t, remain=", scalar @part;
}

say "result2: ", scalar @part;
